implementations/utils/metric.py

```diff
@@ -389,9 +397,45 @@ def is_valid_entry() -> bool:
   return True
 
 
-# TODO(ranran):
-# 1) handle job status
-# 2) handle Airflow retry to avoid duplicate records in tables
+def get_job_status(benchmark_id: str) -> bigquery.JobStatus:
+  """Get job status for the run.
+
+  MISSED - if any failure occurs in initialize & create_queued_resource
+  FAILED - if any failure occurs in setup & run_model (including timeout of
+  run_model)
+  SUCCESS - end-to-end model tests are successful from provision to run_model
+  """
+  context = get_current_context()
+  execution_date = context["dag_run"].logical_date
+  current_dag = context["dag"]
+
+  # check setup status to see if provision step is successful
+  setup_task = current_dag.get_task(task_id=f"{benchmark_id}.provision.setup")
+  setup_ti = TaskInstance(setup_task, execution_date)
+  setup_state = setup_ti.current_state()
+  if setup_state == TaskState.UPSTREAM_FAILED.value:
+    print("The setup state is upstream_failed, and the job status is missed.")
+    return bigquery.JobStatus.MISSED
+
+  # check setup status to see if setup step is successful
+  if setup_state == TaskState.FAILED.value:
+    print("The setup state is failed, and the job status is failed.")
```

Comment:
nit: we should use a logging library like `logging` or `absl.logging`

Reply:
Good call. I tried `absl.logging` previously, and did not see those logs in airflow. So I have a couple of places writing with `print()`. I will have a separate CL to replace all of them.